### PR TITLE
#163426524 Implement social authentication

### PR DESCRIPTION
**Pull request by divinediscipline**
#### What does this PR do?
Implements social authentication for users

#### Description of Task to be completed?
add Google authentication 
add Facebook authentication
add Twitter authentication

#### How should this be manually tested?
Clone the repo, cd into it, run npm install to install all dependencies. Check the env.sample file and input the required api keys and secret. Run `npm run dev` to start the server on port 3000. Navigate to any of the following routes below to sign up or log in with your social account.

- GET /auth/google
- GET /auth/facebook
- GET /auth/twitter

**With test suite:**
- Run `npm test`

#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?
#163426524

#### Screenshots
screen shot 2019-02-11 at 2 50 54 pm
screen shot 2019-02-11 at 2 51 55 pm

### CODE REVIEW

**Review by ozone4real — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Comment by EfeAgare** [comment]
I just notice in  the PR , there is no endpoint defined in **Description of Task to be completed?**. Look into that.

File: server/controllers/userController.js [hunk 1]
```diff
@@ -270,6 +270,41 @@ class UserController extends BaseHelper {
...
+   * authenticated user's data into the database
+   * @param {string} accessToken
+   * @param {string} refreshToken
+   * @param {string} profile
+   * @param {function} done
+   *
+   * @memberOf UserController class
+   */
+  static async handleSocialAuth(accessToken, refreshToken, profile, done) {
+    // console.log('***profile***', profile);
```
Comment (primuse) [inline server/controllers/userController.js]: please take out the commented line

File: package.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "NODE_ENV=test npm run migrate:refresh && cross-env NODE_ENV=test nyc mocha --require @babel/register --require @babel/polyfill ./server/tests/**/*.test.js --timeout 6000 --exit",
+    "test": "NODE_ENV=test npm run migrate:refresh && cross-env NODE_ENV=test nyc mocha --require @babel/register --require @babel/polyfill ./server/tests/**/*.test.js --timeout 10000 --exit",
```
Comment (omenkish) [inline package.json]: Is there a specific reason for this?

File: package.json [hunk 3]
```diff
@@ -21,6 +21,13 @@
   "license": "MIT",
   "dependencies": {
     "@sendgrid/mail": "^6.3.1",
+    "@babel/cli": "^7.2.3",
```
Comment (omenkish) [inline package.json]: We moved all babel dependencies to devDependencies. Please adjust them

File: package.json [hunk 4]
```diff
@@ -21,6 +21,13 @@
...
+    "@babel/core": "^7.2.2",
+    "@babel/polyfill": "^7.2.5",
+    "@babel/preset-env": "^7.3.1",
+    "@babel/register": "^7.0.0",
+    "babel-core": "^7.0.0-bridge.0",
```
Comment (omenkish) [inline package.json]: this has been removed

File: server/controllers/userController.js [hunk 5]
```diff
@@ -258,7 +256,9 @@ class UserController extends BaseHelper {
             returning: true
           }
         ).then(([updatedRows, [updatedUser]]) => {
-          UserController.checkIfDataExist(req, res, updatedRows, { message: 'User not found' });
+          if (!updatedRows) {
```
Comment (omenkish) [inline server/controllers/userController.js:261]: please, why was this changed?

File: server/controllers/userController.js [hunk 6]
```diff
@@ -258,7 +256,9 @@ class UserController extends BaseHelper {
...
+            return errorMessage(res, 404, 'User not found');
```
Comment (omenkish) [inline server/controllers/userController.js:262]: The previous implementation covered for this already.

File: server/db/config/passportConfigs/facebookStrategy.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+dotenv.config();
+
+const FacebookStrategy = require('passport-facebook').Strategy;
+
+const facebook = (passport) => {
+  passport.use(new FacebookStrategy(
+    {
+      clientID: process.env.FACEBOOK_CLIENT_ID,
+      clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
+      callbackURL: 'http://localhost:5000/auth/facebook/callback',
```
Comment (primuse) [inline server/db/config/passportConfigs/facebookStrategy.js]: Why did you change the port from 3000 to 5000? Please change it back

File: server/index.js [hunk 8]
```diff
@@ -18,6 +28,17 @@ app.get('/', (req, res) => res.status(200).json({
...
+app.use(session({
+  secret: 'secret',
+  resave: true,
+  saveUninitialized: false
+}));
+
+app.use(passport.initialize());
+app.use(passport.session());
+
+// passport.use(google);
```
Comment (primuse) [inline server/index.js]: Please take out the commented line

File: server/index.js [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
+import twitter from './db/config/passportConfigs/twitterStrategy';
+
+google(passport);
+facebook(passport);
+twitter(passport);
 
 // Create global app object
 const app = express();
-const PORT = process.env.PORT || 3000;
+const PORT = process.env.PORT || 5000;
```
Comment (omenkish) [inline server/index.js]: Is there a specific reason for changing the port number?

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by biolabalo — DISMISSED** [review]

File: server/controllers/userController.js [hunk 10]
```diff
@@ -285,7 +319,9 @@ class UserController extends BaseHelper {
         where: { email },
         returning: true
       });
-      UserController.checkIfDataExist(req, res, dbUser, { message: 'This email is not registered in our system' });
+      UserController.checkIfDataExist(req, res, dbUser, {
+        message: 'This email is not registered in our system'
+      });
       const { id, username } = dbUser;
       // define token payload and duration
       const jwtKey = process.env.JWT_SECRET;
```
Comment (p-funky) [inline server/controllers/userController.js:329]: How would this be set when you've removed the reference to dotenv?

File: package.json (hunk 2 above)
Comment (p-funky) [reply package.json]: +1

File: server/tests/app/auth.test.js [hunk 11]
```diff
@@ -1,6 +1,23 @@
...
+  id: '108288014452561062871',
+  displayName: 'Divinelove Chukwuemeka',
+  emails: [{ value: 'divinelove.chukwuemeka@andela.com', type: 'account' }],
+  photos: [
+    {
+      value:
+        'https://lh6.googleusercontent.com/-oI45jeLioco/AAAAAAAAAAI/AAAAAAAAAAA/ACevoQNPvTdoQs-JQ7dFLjzGy7erCjHlLw/mo/photo.jpg?sz=50'
+    }
+  ]
+};
```
Comment (p-funky) [inline server/tests/app/auth.test.js:20]: This data seems like your legit actual data. Is it not possible to use some fake data?

File: package.json (hunk 2 above)
Comment (divinediscipline) [reply package.json]: No. Not anymore.

**Review by p-funky — APPROVED** [review]

**Review by ozone4real — APPROVED** [review]
LGTM

**Review by omenkish — APPROVED** [review]
